source/initial_topography_model/interface.cc

```diff
@@ -0,0 +1,173 @@
+    declare_parameters (dealii::ParameterHandler &)
+    {}
+
+
+    template <int dim>
+    void
+    Interface<dim>::parse_parameters (dealii::ParameterHandler &)
+    {}
+
+
+// -------------------------------- Deal with registering initial topography models and automating
+// -------------------------------- their setup and selection at run time
+
+    namespace
+    {
+      std_cxx11::tuple
+      <void *,
+      void *,
+      internal::Plugins::PluginList<Interface<2> >,
+      internal::Plugins::PluginList<Interface<3> > > registered_plugins;
+    }
+
+
+
+    template <int dim>
+    void
+    register_initial_topography_model (const std::string &name,
+                                       const std::string &description,
+                                       void (*declare_parameters_function) (ParameterHandler &),
+                                       Interface<dim> *(*factory_function) ())
+    {
+      std_cxx11::get<dim>(registered_plugins).register_plugin (name,
+                                                               description,
+                                                               declare_parameters_function,
+                                                               factory_function);
+    }
+
+
+    template <int dim>
+    Interface<dim> *
+    create_initial_topography_model (ParameterHandler &prm)
+    {
+      std::string model_name;
+      prm.enter_subsection ("Initial topography model");
+      {
+        model_name = prm.get ("Model name");
+      }
+      prm.leave_subsection ();
+
+      // If one sets the model name to an empty string in the input file,
+      // ParameterHandler produces an error while reading the file. However,
+      // if one omits specifying any model name at all (not even setting it to
+      // the empty string) then the value we get here is the empty string. If
+      // we don't catch this case here, we end up with awkward downstream
+      // errors because the value obviously does not conform to the Pattern.
+      AssertThrow(model_name != "unspecified",
+                  ExcMessage("You need to select a Initial topography model "
+                             "('set Model name' in 'subsection Initial topography model')."));
+
+      return std_cxx11::get<dim>(registered_plugins).create_plugin (model_name,
+                                                                    "Initial topography model::model name",
+                                                                    prm);
+    }
+
+
+
+    template <int dim>
+    void
+    declare_parameters (ParameterHandler &prm)
+    {
+      // declare the entry in the parameter file
+      prm.enter_subsection ("Initial topography model");
+      {
+        const std::string pattern_of_names
+          = std_cxx11::get<dim>(registered_plugins).get_pattern_of_names ();
+        prm.declare_entry ("Model name", "none",
+                           Patterns::Selection (pattern_of_names+"|unspecified"),
```

Comment:
I don't think you need to add `|unspecified`. In that case, you also don't need the `AssertThrow` in the previous function.